internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,351 @@
+/*
+Copyright 2021 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+	"github.com/kube-storage/spec/lib/go/replication"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//  imageMirroringMode is used to indicate the mirroring mode for an RBD image.
+type imageMirroringMode string
+
+const (
+	// imageMirrorModeJournal uses journaling to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeJournal imageMirroringMode = "journal"
+	// imageMirrorModeSnapshot uses snapshots to propagate RBD images between
+	// ceph clusters.
+	imageMirrorModeSnapshot imageMirroringMode = "snapshot"
+)
+
```

Comment:
I would prefer a type for the ReplicationController, if that is possible, like:

```go
type ReplicationController struct {
    ControllerServer
}
```

then all methods in this file can use

```go
func (rc *ReplicationController) EnableReplication(...) {}
```

This helps with splitting the responsibilities of the different controllers, and makes a cleaner API.

It would even be better to have a separate package for each server component, and an internally consumable RbdVolume API. But things are very messy at the moment, so it is not something that can easily be done now. Please consider preparing for this change in the (near) future.

Reply:
Yes, will try to do it. Later as an enhancement :+1:

Reply:
Please mark this with a `TODO:` comment.

Reply:
done.